add basic support for Travis CI

ordex opened the pull request:
Add a .travis.yml file with related scripts in order to
have a special branch built by travis-ci after each push.

At the moment travis-ci has been configured to build for
linux and macos using gcc-5 and clang-5, with both mbedtls
and openssl.

Signed-off-by: Antonio Quartulli <antonio@openvpn.net>

chipitsine commented:
why build lz4 from source ?

ordex commented:
I tried using the system lib but was too old. So I went for building from the source rather than finding a way to tell trusty to install a newer one.

chipitsine commented:
well, we can use either apt-add-repository ... xenial or use docker, but I agree that building from source is also ok

ordex commented:
I tried to stick to the same approach taken by the openvpn repository (i.e. for lzo)

syzzer commented on .travis/build-deps.sh:
```diff
@@ -0,0 +1,132 @@
+        if [ ! -L asio ]; then
+	    rm -Rf asio
+            ln -s download-cache/asio asio
+	fi
```
Indenting looks off

ordex replied:
something was wrong with my vim style file. will fix it. Thanks!

syzzer commented on .travis/build-check.sh:
```diff
@@ -0,0 +1,54 @@
+
+LIBS="${SSL_LIBS} -llz4"
+
+# compiling proto.cpp is enough. no need to compile the client too
```
Why not compile this too?  Is it automatically compiled when building proto.cpp?  Otherwise just build it, so we'll catch compile errors here too.

ordex replied:
no it is not automatically compiled with proto.cpp. ovpncli.cpp is just a stub client interface. But you are right. Better catching errors here too. It's not much effort on our side after all

syzzer commented on .travis/build-deps.sh:
```diff
@@ -0,0 +1,132 @@
+        (
+            cd "lz4-${LZ4_VERSION}"
+            make default CC=$CC CXX=$CXX
+	    make install PREFIX="${PREFIX}"
```
More indenting

syzzer commented on .travis.yml:
```diff
@@ -0,0 +1,77 @@
+sudo: required
```
Do you actually need sudo?  We need it in 2.0 for installing mingw (because the version in the standard repo is too old), but you don't seem to need it (yet). If you don't need it, don't specify it. Your builds will then be faster (there is only a limited number of 'sudo' slaves available).

ordex replied:
Oh ok. Thank for the hint. I wasn't sure about it either, but I did not investigate further. I'll remove it.

syzzer commented on .travis.yml:
```diff
@@ -0,0 +1,77 @@
+
+matrix:
+  include:
+    - env: SSLLIB="openssl" CC="gcc-5" CXX="g++-5"
```
Why specifically gcc/clang 5?

ordex replied:
to get full support for C++14

ordex commented:
- removed constraint on branch to build (now build any branch)
- removed sudo requirement
- fixed style (i.e. tabs converted to spaces, make the scripts consistent)
- compile also ovpncli

ordex commented:
weird. the osx build is now failing. Never seen such error...I'll investigate

ordex commented:
I'll also release the constraint on C++ and reduce to C++11 as discussed.

ordex commented:
@mattock I have fixed the latest glitches on osx. it now build fine. You can have another look and then merge if you think it's fine!
